include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -0,0 +1,111 @@
+#ifndef _NBL_BUILTIN_HLSL_WORKGROUP_FFT_INCLUDED_
+#define _NBL_BUILTIN_HLSL_WORKGROUP_FFT_INCLUDED_
+
+#include <nbl/builtin/hlsl/subgroup/fft.hlsl>
+#include <nbl/builtin/hlsl/bit.hlsl>
+
+namespace nbl 
+{
+namespace hlsl
+{
+namespace workgroup
+{
+
+template<typename Accessor, typename Scalar, uint32_t WorkgroupSize>
+void store(NBL_REF_ARG(Accessor) sharedmemAccessor, uint32_t threadID, NBL_CONST_REF_ARG(complex_t<Scalar>) lo, NBL_CONST_REF_ARG(complex_t<Scalar>) hi) {
+    sharedmemAccessor.set(threadID + 0 * WorkgroupSize, lo.real());
+    sharedmemAccessor.set(threadID + 1 * WorkgroupSize, hi.real());
+    sharedmemAccessor.set(threadID + 2 * WorkgroupSize, lo.imag());  
+    sharedmemAccessor.set(threadID + 3 * WorkgroupSize, hi.imag());    
+}
+
+template<typename Accessor, typename Scalar, uint32_t WorkgroupSize>
+void load(NBL_CONST_REF_ARG(Accessor) sharedmemAccessor, uint32_t threadID, NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi) {
+    lo.real(sharedmemAccessor.get(threadID + 0 * WorkgroupSize));
+    hi.real(sharedmemAccessor.get(threadID + 1 * WorkgroupSize));
+    lo.imag(sharedmemAccessor.get(threadID + 2 * WorkgroupSize));  
+    hi.imag(sharedmemAccessor.get(threadID + 3 * WorkgroupSize));    
+}
+
+template<typename Accessor, typename Scalar, uint32_t WorkgroupSize, bool inverse>
+void FFT_loop(NBL_REF_ARG(Accessor) sharedmemAccessor, uint32_t step, uint32_t threadID){
+    const uint32_t sub_ix = threadID & (step - 1);
+    const uint32_t lo_x_ix = impl::bitfieldInsert(threadID, 0u ,sub_ix, 1u);
+    const uint32_t hi_x_ix = lo_x_ix | step;
+    const uint32_t lo_y_ix = lo_x_ix + 2 * WorkgroupSize;
+    const uint32_t hi_y_ix = hi_x_ix + 2 * WorkgroupSize;
+
+    sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
+    complex_t<Scalar> lo = {sharedmemAccessor.get(lo_x_ix), sharedmemAccessor.get(lo_y_ix)};
+    complex_t<Scalar> hi = {sharedmemAccessor.get(hi_x_ix), sharedmemAccessor.get(hi_y_ix)};
+
+    fft::DIX<Scalar, inverse>::radix2(fft::twiddle<Scalar, inverse>(sub_ix, step << 1), lo, hi);
+
+    // Share results between threads
+    sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
+    sharedmemAccessor.set(lo_x_ix, lo.real());
+    sharedmemAccessor.set(hi_x_ix, hi.real());
+    sharedmemAccessor.set(lo_y_ix, lo.imag());
+    sharedmemAccessor.set(hi_y_ix, hi.imag());     
+}
+
+// When doing a 1D FFT of size N, we call N/2 threads to do the work, where each thread of index 0 <= k < N/2 is in charge of computing two elements
+// in the output array, those of index k and k + N/2. The first is "lo" and the latter is "hi"
+template<typename Accessor, typename Scalar, uint32_t WorkgroupSize, uint32_t SubgroupSize, bool inverse>
+void FFT(NBL_REF_ARG(Accessor) sharedmemAccessor, NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi){
```

Comment:
instead of taking `lo` and `hi` take a `preloaded[K]` array where `K=N/WorkgroupSize` and must be PoT (for now, we can experiment later with specializations mixing radices to get K=3,5,6,7)

Reply:
you also can't template on the SubgroupSize :(

Reply:
also you need two accessors:
- SharedMemoryAccessor (for workgroup FFT)
- Accessor (for getting/setting outputs)

Both you shall wrap in  `MemoryAdaptor` to turn `uint32_t` gets/sets into `vector<Scalar>` sets and gets

Reply:
I'd do this
```cpp
template<uint16_t K, bool Inverse, class device_capabilities=void>
struct FFT;

// 2 items per invocation forward specialization
template<class device_capabilities>
struct FFT<2,false,device_capabilities>
{
    template<typename Accessor, typename ShaderMemoryAccessor>
    static void __call(NBL_REF_ARG(Accessor) accessor, NBL_REF_ARG(ShaderMemoryAccessor) sharedmemAccessor);
};
// 2 items per invocation inverse specialization
template<class device_capabilities>
struct FFT<2,true,device_capabilities>
{
    template<typename Accessor, typename ShaderMemoryAccessor>
    static void __call(NBL_REF_ARG(Accessor) accessor, NBL_REF_ARG(ShaderMemoryAccessor) sharedmemAccessor);
};

// then define 4,8,16 in terms of calling the FFT<2> and doing the special radix steps before/after
template<uint16_t K, bool Inverse, class device_capabilities>
struct FFT
{
    template<typename Accessor, typename ShaderMemoryAccessor>
    static void __call(NBL_REF_ARG(Accessor) accessor, NBL_REF_ARG(ShaderMemoryAccessor) sharedmemAccessor)
    {
        if (!Inverse)
        {
           ... special steps ...
        }
        FFT<2,Inverse,device_capabilities>::template __call<Accessor,SharedMemoryAccessor>(access,sharedMemAccessor);
        if (Inverse)
        {
           ... special steps ...
        }
    }
};
```